libafl/src/corpus/probabilistic_sampling.rs

```diff
@@ -0,0 +1,197 @@
+where
+    I: Input,
+    S: HasCorpus<I> + HasMetadata + HasRand,
+    F: FavFactor<I>,
+{
+    phantom: PhantomData<(I, S, F)>,
+}
+
+/// A state metadata holding a map of probability of corpus elements.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ProbabilityMetadata {
+    /// corpus index -> probability
+    pub map: HashMap<usize, f64>,
+    /// total probability of all items in the map
+    pub total_probability: f64,
+}
+
+crate::impl_serdeany!(ProbabilityMetadata);
+
+impl ProbabilityMetadata {
+    /// Creates a new [`struct@ProbabilityMetadata`]
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            map: HashMap::default(),
+            total_probability: 0.0,
+        }
+    }
+}
+
+impl Default for ProbabilityMetadata {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<I, S, F> ProbabilitySamplingCorpusScheduler<I, S, F>
+where
+    I: Input,
+    S: HasCorpus<I> + HasMetadata + HasRand,
+    F: FavFactor<I>,
+{
+    /// Creates a new [`struct@ProbabilitySamplingCorpusScheduler`]
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            phantom: PhantomData,
+        }
+    }
+
+    /// Calculate the score and store in `ProbabilityMetadata`
+    #[allow(clippy::cast_precision_loss)]
+    #[allow(clippy::unused_self)]
+    pub fn store_probability(&self, state: &mut S, idx: usize) -> Result<(), Error> {
+        let factor = F::compute(&mut *state.corpus().get(idx)?.borrow_mut())?;
+        if factor == 0 {
+            return Err(Error::DivByZero(
+                "Infinity probability calculated for probabilistic sampling scheduler".into(),
+            ));
+        }
+        let meta = state
+            .metadata_mut()
+            .get_mut::<ProbabilityMetadata>()
+            .unwrap();
+        let prob = 1.0 / (factor as f64);
+        meta.map.insert(idx, prob);
+        meta.total_probability += prob;
+        Ok(())
+    }
+}
+
+impl<I, S, F> CorpusScheduler<I, S> for ProbabilitySamplingCorpusScheduler<I, S, F>
+where
+    I: Input,
+    S: HasCorpus<I> + HasMetadata + HasRand,
+    F: FavFactor<I>,
+{
+    fn on_add(&self, state: &mut S, idx: usize) -> Result<(), Error> {
+        if state.metadata().get::<ProbabilityMetadata>().is_none() {
+            state.add_metadata(ProbabilityMetadata::new());
+        }
+        self.store_probability(state, idx)
+    }
+
+    /// Gets the next entry
+    #[allow(clippy::cast_precision_loss)]
+    fn next(&self, state: &mut S) -> Result<usize, Error> {
+        if state.corpus().count() == 0 {
+            Err(Error::Empty(String::from("No entries in corpus")))
+        } else {
+            let rand_prob: f64 = (state.rand_mut().below(100) as f64) / 100.0;
+            let meta = state.metadata().get::<ProbabilityMetadata>().unwrap();
+            let threshold = meta.total_probability * rand_prob;
+            let mut k: f64 = 0.0;
+            for (idx, prob) in meta.map.iter() {
+                k += prob;
```

Comment:
maybe you can compute the total prob up until `idx` in `add()` to improve the efficiency 😀  and then you don't have to do this addition in `next()`.
because I think next() called far more often than `add()`
but it already LGTM

Reply:
you can this one.
`BTreeMap`: https://doc.rust-lang.org/stable/alloc/collections/btree_map/struct.BTreeMap.html